### PR TITLE
Added ltp testcase which runs runltplite.sh as default

### PR DESCRIPTION
**Pull request by santhosh93**
This script runs runltplite.sh as a default one and it also runs runltp when run with multiplex option.
Currently the script fails when run with multiplex option since the below PR has to be merged in avocado.
PR: 
https://github.com/avocado-framework/avocado/pull/1241

### CODE REVIEW

**Comment by sathnaga** [comment]
ACKed

File: generic/ltp.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        os.mkdir(ltpbin_dir)
+        process.system('./configure --prefix=%s' % ltpbin_dir)
+        build.make(self.srcdir, extra_args='-j %d' %
+                   multiprocessing.cpu_count())
+        build.make(self.srcdir, extra_args='install')
+
+    def test(self):
+        script = self.params.get('script', default='runltplite.sh')
+        args = self.params.get('args', default=' ')
+        if script == 'runltp':
```
Comment (sathnaga) [inline generic/ltp.py:57]: default can be empty string, minor fix

File: generic/ltp.py.data/ltp.yaml [hunk 2]
```diff
@@ -0,0 +1,58 @@
+setup:
+    general: !mux
+        default:
+            script: 'runltplite.sh'
+            args: ' '
```
Comment (sathnaga) [inline generic/ltp.py.data/ltp.yaml:5]: args can be empty string

File: generic/ltp.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+            logfile = os.path.join(self.logdir, 'ltp.log')
```
Comment (ldoktor) [inline generic/ltp.py:58]: the `self.outputdir` would be probably more appropriate for additional files. The `logdir` itself has defined hierarchy and no other files should be created there (theoretically).

File: generic/ltp.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            failcmdfile = os.path.join(self.logdir, 'failcmdfile')
+            skipfile = os.path.join(self.datadir, 'skipfile')
+            args2 = '-q -p -l %s -C %s -d %s -S %s' % (
```
Comment (ldoktor) [inline generic/ltp.py:61]: nitpick, how about:

``` python
args += ("-q -p -l %s -C %s -d %s -S %s"
         % (logfile, failcmdfile, self.srcdir, skipfile))
```

:wink:

File: generic/ltp.py (hunk 1 above)
Comment (ldoktor) [inline generic/ltp.py:57]: in deed, empty string would be better.

File: generic/ltp.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                logfile, failcmdfile, self.srcdir, skipfile)
+            args = args + ' ' + args2
+        ltpbin_dir = os.path.join(self.srcdir, 'bin')
+        cmd = os.path.join(ltpbin_dir, script) + ' ' + args
+        result = process.run(cmd, ignore_status=True)
+        failed_tests = []
```
Comment (ldoktor) [inline generic/ltp.py:67]: There are nice comments in the original script, do you have any specific reasons for not including them here?

File: generic/ltp.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    def setUp(self):
+        sm = SoftwareManager()
+        detected_distro = distro.detect()
+        deps = ['gcc', 'git', 'make', 'automake', 'autoconf']
+        for package in deps:
+            if package == 'git' and detected_distro.name == "SuSE":
+                package = 'git-core'
+            if not sm.check_installed(package) and not sm.install(package):
+                self.error(package + ' is needed for the test to be run')
+        git.get_repo('https://github.com/linux-test-project/ltp.git',
```
Comment (ldoktor) [inline generic/ltp.py:43]: It'd be much better to use `fetch_asset` as github allows one to download master as zip file. Anyway I can take care of that after merging this version...

File: generic/ltp.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            skipfile = os.path.join(self.datadir, 'skipfile')
+            args2 = '-q -p -l %s -C %s -d %s -S %s' % (
+                logfile, failcmdfile, self.srcdir, skipfile)
+            args = args + ' ' + args2
+        ltpbin_dir = os.path.join(self.srcdir, 'bin')
+        cmd = os.path.join(ltpbin_dir, script) + ' ' + args
+        result = process.run(cmd, ignore_status=True)
+        failed_tests = []
+        for line in result.stdout.splitlines():
+            if set(('TFAIL', 'TBROK', 'TWARN')).intersection(line.split()):
```
Comment (ldoktor) [inline generic/ltp.py:69]: this set could be created only once before the for loop...

File: generic/ltp.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+                test_name = line.strip().split(' ')[0]
+                if test_name not in failed_tests:
+                    failed_tests.append(test_name)
+
+        if failed_tests:
+            self.error("LTP tests failed: %s" % failed_tests)
```
Comment (ldoktor) [inline generic/ltp.py:75]: How about: `self.error("Some LTP tests failed: %s", ", ".join(failed_tests))`?

File: generic/ltp.py.data/ltp.yaml (hunk 2 above)
Comment (ldoktor) [inline generic/ltp.py.data/ltp.yaml:5]: or None

**Comment by ldoktor** [comment]
Hello @santhosh93, there are few nitpicks, but nothing major. I'll try to run the test suite and if everything behaves correctly, I'll merge this and add the tweaks myself.

File: generic/ltp.py.data/ltp.yaml [hunk 9]
```diff
@@ -0,0 +1,58 @@
+setup:
```
Comment (ldoktor) [inline generic/ltp.py.data/ltp.yaml:1]: Btw I'm wondering why those prefixes. I'd simply put all params in global so one could inject this file anywhere...

Comment (ldoktor) [inline generic/ltp.py.data/ltp.yaml:1]: I mean I'd simply use:

``` yaml
!mux
default:
    script: ...
runltp: !mux
    script: ...
```

**Comment by ldoktor** [comment]
It seems this does not cover all possible failures. I tried the NUMA test on non-NUMA machine and got this report:

```
2016-06-08 18:23:28,453 multiplexer      L0182 DEBUG| PARAMS (key=args, path=*, default= ) => '-f numa'
2016-06-08 18:23:28,454 process          L0331 INFO | Running '/var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/bin/runltp -f numa -q -p -l /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/ltp.log -C /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/failcmdfile -d /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src -S /home/medic/Work/Projekty/avocado/avocado-misc-tests/generic/ltp.py.data/skipfile'
2016-06-08 18:23:28,466 process          L0418 DEBUG| [stdout] INFO: creating /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/bin/output directory
2016-06-08 18:23:28,473 process          L0418 DEBUG| [stdout] Checking for required user/group ids
2016-06-08 18:23:28,473 process          L0418 DEBUG| [stdout] 
2016-06-08 18:23:28,478 process          L0418 DEBUG| [stdout] 'nobody' user id and group found.
2016-06-08 18:23:28,479 process          L0418 DEBUG| [stdout] 'bin' user id and group found.
2016-06-08 18:23:28,479 process          L0418 DEBUG| [stdout] 'daemon' user id and group found.
2016-06-08 18:23:28,479 process          L0418 DEBUG| [stdout] Users group found.
2016-06-08 18:23:28,479 process          L0418 DEBUG| [stdout] Sys group found.
2016-06-08 18:23:28,485 process          L0418 DEBUG| [stdout] Required users/groups exist.
2016-06-08 18:23:28,612 process          L0418 DEBUG| [stderr] losetup: /dev/loop0: zařízení loop se nepodařilo nastavit: Operace zamítnuta
2016-06-08 18:23:28,613 process          L0418 DEBUG| [stdout] losetup failed to create block device
2016-06-08 18:23:28,613 process          L0418 DEBUG| [stdout] no block device was specified on commandline.
2016-06-08 18:23:28,613 process          L0418 DEBUG| [stdout] Block device could not be created using loopback device
2016-06-08 18:23:28,613 process          L0418 DEBUG| [stdout] Tests which require block device are disabled.
2016-06-08 18:23:28,614 process          L0418 DEBUG| [stdout] You can specify it with option -b
2016-06-08 18:23:28,614 process          L0418 DEBUG| [stdout] no big block device was specified on commandline.
2016-06-08 18:23:28,614 process          L0418 DEBUG| [stdout] Tests which require a big block device are disabled.
2016-06-08 18:23:28,614 process          L0418 DEBUG| [stdout] You can specify it with option -z
2016-06-08 18:23:28,615 process          L0418 DEBUG| [stdout] INFO: Test start time: St čen  8 18:23:28 CEST 2016
2016-06-08 18:23:28,615 process          L0418 DEBUG| [stdout] COMMAND:    /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/bin/bin/ltp-pan  -q  -e -S   -a 798     -n 798  -p  -f /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/ltp-xkBMCnFog9/alltests -l /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/ltp.log  -C /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/failcmdfile -T /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/bin/output/LTP_RUN_ON-ltp.log.tconf
2016-06-08 18:23:28,615 process          L0418 DEBUG| [stdout] LOG File: /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/ltp.log
2016-06-08 18:23:28,615 process          L0418 DEBUG| [stdout] FAILED COMMAND File: /home/medic/avocado/job-results/job-2016-06-08T18.21-6a4a6f4/test-results/1-generic_ltp.py:ltp.test/failcmdfile
2016-06-08 18:23:28,617 process          L0418 DEBUG| [stdout] TCONF COMMAND File: /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/bin/output/LTP_RUN_ON-ltp.log.tconf
2016-06-08 18:23:28,617 process          L0418 DEBUG| [stdout] Running tests.......
2016-06-08 18:23:28,629 process          L0418 DEBUG| [stdout] Initnuma    0  TINFO  :  INIT: Numa tests will start now !!
2016-06-08 18:23:28,631 process          L0418 DEBUG| [stdout] Initnuma    0  TINFO  :  The system contains 1 nodes:  0 
2016-06-08 18:23:28,632 process          L0418 DEBUG| [stdout] Initnuma    1  TCONF  :  ltpapicmd.c:200: your machine does not support numa policy or
2016-06-08 18:23:28,632 process          L0418 DEBUG| [stdout]                         your machine is not a NUMA machine
2016-06-08 18:23:28,633 process          L0418 DEBUG| [stdout] exitnuma    0  TINFO  :  CLEAN: removing /var/tmp/avocado_suO49D/1-generic_ltp.py_ltp.test/src/ltp-xkBMCnFog9/tst_numa.847
2016-06-08 18:23:28,635 process          L0418 DEBUG| [stdout] move_pages01    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,635 process          L0418 DEBUG| [stdout] move_pages01    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,637 process          L0418 DEBUG| [stdout] move_pages02    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,637 process          L0418 DEBUG| [stdout] move_pages02    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,639 process          L0418 DEBUG| [stdout] chown: změnění vlastnictví souboru „move_pages03“: Operace není povolena
2016-06-08 18:23:28,640 process          L0418 DEBUG| [stdout] move_pages04    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,641 process          L0418 DEBUG| [stdout] move_pages04    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,642 process          L0418 DEBUG| [stdout] move_pages05    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,642 process          L0418 DEBUG| [stdout] move_pages05    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,644 process          L0418 DEBUG| [stdout] move_pages06    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,644 process          L0418 DEBUG| [stdout] move_pages06    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,646 process          L0418 DEBUG| [stdout] move_pages07    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,646 process          L0418 DEBUG| [stdout] move_pages07    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,648 process          L0418 DEBUG| [stdout] move_pages08    1  TCONF  :  move_pages08.c:138: move_pages: E2BIG was removed in commit 3140a227
2016-06-08 18:23:28,648 process          L0418 DEBUG| [stdout] move_pages08    2  TCONF  :  move_pages08.c:138: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,649 process          L0418 DEBUG| [stdout] move_pages09    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,650 process          L0418 DEBUG| [stdout] move_pages09    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,651 process          L0418 DEBUG| [stdout] move_pages10    1  TCONF  :  move_pages_support.c:412: NUMA support not provided
2016-06-08 18:23:28,651 process          L0418 DEBUG| [stdout] move_pages10    2  TCONF  :  move_pages_support.c:412: Remaining cases not appropriate for configuration
2016-06-08 18:23:28,653 process          L0418 DEBUG| [stdout] chown: změnění vlastnictví souboru „move_pages11“: Operace není povolena
2016-06-08 18:23:28,654 process          L0418 DEBUG| [stdout] INFO: ltp-pan reported some tests FAIL
2016-06-08 18:23:28,654 process          L0418 DEBUG| [stdout] LTP Version: 20160510-64-ge0bfa7d
2016-06-08 18:23:28,654 process          L0418 DEBUG| [stdout] INFO: Test end time: St čen  8 18:23:28 CEST 2016
2016-06-08 18:23:28,655 process          L0418 DEBUG| [stdout] 
2016-06-08 18:23:28,655 process          L0418 DEBUG| [stdout]        ###############################################################
2016-06-08 18:23:28,655 process          L0418 DEBUG| [stdout] 
2016-06-08 18:23:28,656 process          L0418 DEBUG| [stdout]             Done executing testcases.
2016-06-08 18:23:28,656 process          L0418 DEBUG| [stdout]             LTP Version:  20160510-64-ge0bfa7d
2016-06-08 18:23:28,656 process          L0418 DEBUG| [stdout]        ###############################################################
2016-06-08 18:23:28,656 process          L0418 DEBUG| [stdout] 
2016-06-08 18:23:28,656 process          L0418 DEBUG| [stderr] losetup: /dev/loop0: odpojení selhalo: Operace zamítnuta
```

But the overall test status in avocado was `PASS`.

Anyway the rest is behaving correctly, so I'm merging this test and we can fix the little issues later. I'll take care of the mentioned ones as well as this non-NUMA false-positive. I'll notify you when pushing the changes.

**Comment by santhosh93** [comment]
@ldoktor : Thank you for the comments and merging the code.

**Comment by ldoktor** [comment]
Thank _you_ for contribution. Would you please review the https://github.com/avocado-framework/avocado-misc-tests/pull/52 I'll take a look at the symlink problem, but at least the first 2 patches are ready for merge, unless they break something for you. Hopefully I'll workaround/fix the symlink issue, which would speedup the testing significantly as it'd cache the LTP.

**Pull request merged**